ModelViewIntentSample/search/src/main/java/com/novoda/movies/mvi/search/presentation/SearchActivity.kt

```diff
@@ -0,0 +1,89 @@
+package com.novoda.movies.mvi.search.presentation
+
+import android.arch.lifecycle.ViewModel
+import android.arch.lifecycle.ViewModelProvider
+import android.arch.lifecycle.ViewModelProviders
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.util.Log
+import android.view.View
+import android.view.View.*
+import com.novoda.movies.mvi.search.ActionProvider
+import com.novoda.movies.mvi.search.Dependencies
+import com.novoda.movies.mvi.search.R
+import com.novoda.movies.mvi.search.ViewRender
+import com.novoda.movies.mvi.search.domain.SearchAction
+import com.novoda.movies.mvi.search.domain.SearchDependencyProvider
+import com.novoda.movies.mvi.search.domain.SearchState
+import com.novoda.movies.mvi.search.domain.SearchState.*
+import io.reactivex.Observable
+import kotlinx.android.synthetic.main.activity_search.*
+
+internal class SearchActivity : AppCompatActivity(),
+        ActionProvider<SearchAction>,
+        ViewRender<SearchState> {
+
+    private lateinit var searchInput: SearchInputView
+    private lateinit var resultsView: SearchResultsView
+
+    private lateinit var viewModel: SearchViewModel
+
+    override val actions: Observable<SearchAction>
+        get() = searchInput.actions
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        viewModel = ViewModelProviders.of(this, SearchViewModelFactory(this))
+                .get(SearchViewModel::class.java)
+        setContentView(R.layout.activity_search)
+        searchInput = search_input
+        resultsView = search_results
+
+        viewModel.wire()
+    }
+
+    override fun onStart() {
+        super.onStart()
+        viewModel.bind(this, this)
+    }
+
+    override fun render(state: SearchState) {
+        when (state) {
+            is Content -> {
+                if (searchInput.currentQuery != state.queryString) {
```

Comment:
why do you do this check on the view?

Reply:
Good question: 

### TL;DR
Unless we are resuming from an activity destroyed, the textInput would be asked to render the exact same text that is already rendering every-time the user is typing something. 
In this case I found that updating the textField too often would result in flashes in the UI and some missing character while typing too fast. 

### Context
The text in the input field is part of the state, in this way we can example restore it after an Activity has been destroyed and recreated. 

That means that for every character introduced in the textInput: 
- An **Action** is emitted _ChangeQuery_
- The middleware emit the relative **Change** _SearchQueryUpdate_
- The Store emits a new **State**
- The view is asked to render the new generated state

In complex UI we want to avoid to render changes that are not necessary or that can compromise the User Experience. 

IDEALLY we should use some UI-Framework that implement the diffing of the changes to have the best of this architecture, as for example
[Litho from Facebook](https://github.com/facebook/litho)
[Domic from Lyft](https://github.com/lyft/domic)

As a simple workaround we can also simply debounce the _ChangeQuery_ emissions so that we don't flood the components until the user has finished typing

Reply:
Yeah this doesn't sound like something the view should control, I would vote for debouncing the emissions 👍 this would allow you to change behaviour in the correct place even when the view is changed to a different one. Right now without me asking this question it feels your above explanation is hidden in the view, rather than shown more where control should happen (the intent?/ChangeQuery)